File: parse/src/main/java/com/parse/ParseRESTCommand.java
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse;

import com.parse.http.ParseHttpBody;
import com.parse.http.ParseHttpRequest;
import com.parse.http.ParseHttpResponse;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONStringer;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;

import bolts.Task;

/**
 * A helper object to send requests to the server.
 */
class ParseRESTCommand extends ParseRequest<JSONObject> {

    /* package */ static final String HEADER_APPLICATION_ID = "X-Parse-Application-Id";
    /* package */ static final String HEADER_CLIENT_KEY = "X-Parse-Client-Key";
    /* package */ static final String HEADER_APP_BUILD_VERSION = "X-Parse-App-Build-Version";
    /* package */ static final String HEADER_APP_DISPLAY_VERSION = "X-Parse-App-Display-Version";
    /* package */ static final String HEADER_OS_VERSION = "X-Parse-OS-Version";

    /* package */ static final String HEADER_INSTALLATION_ID = "X-Parse-Installation-Id";
    /* package */ static final String USER_AGENT = "User-Agent";
    private static final String HEADER_SESSION_TOKEN = "X-Parse-Session-Token";
    private static final String HEADER_MASTER_KEY = "X-Parse-Master-Key";
    private static final String PARAMETER_METHOD_OVERRIDE = "_method";

    // Set via Parse.initialize(Configuration)
    /* package */ static URL server = null;
    /* package */ final JSONObject jsonParameters;
    // Headers
    private final String sessionToken;
    public String masterKey;
    /* package */ String httpPath;
    private String installationId;
    private String operationSetUUID;
    private String localId;
    public ParseRESTCommand(
            String httpPath,
            ParseHttpRequest.Method httpMethod,
            Map<String, ?> parameters,
            String sessionToken) {
        this(
                httpPath,
                httpMethod,
                parameters != null ? (JSONObject) NoObjectsEncoder.get().encode(parameters) : null,
                sessionToken);
    }

    public ParseRESTCommand(
            String httpPath,
            ParseHttpRequest.Method httpMethod,
            JSONObject jsonParameters,
            String sessionToken) {
        this(httpPath, httpMethod, jsonParameters, null, sessionToken);
    }
    private ParseRESTCommand(
            String httpPath,
            ParseHttpRequest.Method httpMethod,
            JSONObject jsonParameters,
            String localId, String sessionToken) {
        super(httpMethod, createUrl(httpPath));

        this.httpPath = httpPath;
        this.jsonParameters = jsonParameters;
        this.localId = localId;
        this.sessionToken = sessionToken;
    }

    /* package */ ParseRESTCommand(Init<?> builder) {
        super(builder.method, createUrl(builder.httpPath));
        sessionToken = builder.sessionToken;
        installationId = builder.installationId;
        masterKey = builder.masterKey;

        httpPath = builder.httpPath;
        jsonParameters = builder.jsonParameters;
        operationSetUUID = builder.operationSetUUID;
        localId = builder.localId;
    }

    private static LocalIdManager getLocalIdManager() {
        return ParseCorePlugins.getInstance().getLocalIdManager();
    }

    public static ParseRESTCommand fromJSONObject(JSONObject jsonObject) {
        String httpPath = jsonObject.optString("httpPath");
        ParseHttpRequest.Method httpMethod =
                ParseHttpRequest.Method.fromString(jsonObject.optString("httpMethod"));
        String sessionToken = jsonObject.optString("sessionToken", null);
        String localId = jsonObject.optString("localId", null);
        JSONObject jsonParameters = jsonObject.optJSONObject("parameters");

        return new ParseRESTCommand(httpPath, httpMethod, jsonParameters, localId, sessionToken);
    }

    private static String createUrl(String httpPath) {
        // We send all parameters for GET/HEAD/DELETE requests in a post body,
        // so no need to worry about query parameters here.
        if (httpPath == null) {
            return server.toString();
        }

        try {
            return new URL(server, httpPath).toString();
        } catch (MalformedURLException ex) {
            throw new RuntimeException(ex);
        }
    }

    // Encodes the object to JSON, but ensures that JSONObjects
    // and nested JSONObjects are encoded with keys in alphabetical order.
    static String toDeterministicString(Object o) throws JSONException {
        JSONStringer stringer = new JSONStringer();
        addToStringer(stringer, o);
        return stringer.toString();
    }

    // Uses the provided JSONStringer to encode this object to JSON, but ensures that JSONObjects and
    // nested JSONObjects are encoded with keys in alphabetical order.
    private static void addToStringer(JSONStringer stringer, Object o) throws JSONException {
        if (o instanceof JSONObject) {
            stringer.object();
            JSONObject object = (JSONObject) o;
            Iterator<String> keyIterator = object.keys();
            ArrayList<String> keys = new ArrayList<>();
            while (keyIterator.hasNext()) {
                keys.add(keyIterator.next());
            }
            Collections.sort(keys);

            for (String key : keys) {
                stringer.key(key);
                addToStringer(stringer, object.opt(key));
            }

            stringer.endObject();
            return;
        }

        if (o instanceof JSONArray) {
            JSONArray array = (JSONArray) o;
            stringer.array();
            for (int i = 0; i < array.length(); ++i) {
                addToStringer(stringer, array.get(i));
            }
            stringer.endArray();
            return;
        }

        stringer.value(o);
    }

    /* package */
    static boolean isValidCommandJSONObject(JSONObject jsonObject) {
        return jsonObject.has("httpPath");
    }

    // This function checks whether a json object is a valid /2 ParseCommand json.
    /* package */
    static boolean isValidOldFormatCommandJSONObject(JSONObject jsonObject) {
        return jsonObject.has("op");
    }

    protected static void getLocalPointersIn(Object container, ArrayList<JSONObject> localPointers)
            throws JSONException {
        if (container instanceof JSONObject) {
            JSONObject object = (JSONObject) container;
            if ("Pointer".equals(object.opt("__type")) && object.has("localId")) {
                localPointers.add((JSONObject) container);
                return;
            }

            Iterator<String> keyIterator = object.keys();
            while (keyIterator.hasNext()) {
                String key = keyIterator.next();
                getLocalPointersIn(object.get(key), localPointers);
            }
        }

        if (container instanceof JSONArray) {
            JSONArray array = (JSONArray) container;
            for (int i = 0; i < array.length(); ++i) {
                getLocalPointersIn(array.get(i), localPointers);
            }
        }
    }

    protected void addAdditionalHeaders(ParseHttpRequest.Builder requestBuilder) {
        if (installationId != null) {
            requestBuilder.addHeader(HEADER_INSTALLATION_ID, installationId);
        }
        if (sessionToken != null) {
            requestBuilder.addHeader(HEADER_SESSION_TOKEN, sessionToken);
        }
        if (masterKey != null) {
            requestBuilder.addHeader(HEADER_MASTER_KEY, masterKey);
        }
    }

    @Override
    protected ParseHttpRequest newRequest(
            ParseHttpRequest.Method method,
            String url,
            ProgressCallback uploadProgressCallback) {
        ParseHttpRequest request;
        if (jsonParameters != null &&
                method != ParseHttpRequest.Method.POST &&
                method != ParseHttpRequest.Method.PUT) {
            // The request URI may be too long to include parameters in the URI.
            // To avoid this problem we send the parameters in a POST request json-encoded body
            // and add a http method override parameter in newBody.
            request = super.newRequest(ParseHttpRequest.Method.POST, url, uploadProgressCallback);
        } else {
            request = super.newRequest(method, url, uploadProgressCallback);
        }
        ParseHttpRequest.Builder requestBuilder = new ParseHttpRequest.Builder(request);
        addAdditionalHeaders(requestBuilder);
        return requestBuilder.build();
    }

    @Override
    protected ParseHttpBody newBody(ProgressCallback uploadProgressCallback) {
        if (jsonParameters == null) {
            String message = String.format("Trying to execute a %s command without body parameters.",
                    method.toString());
            throw new IllegalArgumentException(message);
        }

        try {
            JSONObject parameters = jsonParameters;
            if (method == ParseHttpRequest.Method.GET ||
                    method == ParseHttpRequest.Method.DELETE) {
                // The request URI may be too long to include parameters in the URI.
                // To avoid this problem we send the parameters in a POST request json-encoded body
                // and add a http method override parameter.
                parameters = new JSONObject(jsonParameters.toString());
                parameters.put(PARAMETER_METHOD_OVERRIDE, method.toString());
            }
            return new ParseByteArrayHttpBody(parameters.toString(), "application/json");
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    @Override
    public Task<JSONObject> executeAsync(
            final ParseHttpClient client,
            final ProgressCallback uploadProgressCallback,
            final ProgressCallback downloadProgressCallback,
            final Task<Void> cancellationToken) {
        resolveLocalIds();
        return super.executeAsync(
                client, uploadProgressCallback, downloadProgressCallback, cancellationToken);
    }

    @Override
    protected Task<JSONObject> onResponseAsync(ParseHttpResponse response,
                                               ProgressCallback downloadProgressCallback) {
        String content;
        InputStream responseStream = null;
        try {
            responseStream = response.getContent();
            content = new String(ParseIOUtils.toByteArray(responseStream));
        } catch (IOException e) {
            return Task.forError(e);
        } finally {
            ParseIOUtils.closeQuietly(responseStream);
        }

        // We need to check for errors differently in /1/ than /2/ since object data in /2/ was
        // encapsulated in "data" and everything was 200, but /2/ everything is in the root JSON,
        // but errors are status 4XX.
        // See https://quip.com/4pbbA9HbOPjQ
        int statusCode = response.getStatusCode();
        if (statusCode >= 200 && statusCode < 600) { // Assume 3XX is handled by http library
            JSONObject json;
            try {
                json = new JSONObject(content);

                if (statusCode >= 400 && statusCode < 500) { // 4XX
                    return Task.forError(newPermanentException(json.optInt("code"), json.optString("error")));
                } else if (statusCode >= 500) { // 5XX
                    return Task.forError(newTemporaryException(json.optInt("code"), json.optString("error")));
                }

                return Task.forResult(json);
            } catch (JSONException e) {
                return Task.forError(newTemporaryException("bad json response", e));
            }
        }

        return Task.forError(newPermanentException(ParseException.OTHER_CAUSE, content));
    }

    // Creates a somewhat-readable string that uniquely identifies this command.
    public String getCacheKey() {
        String json;
        if (jsonParameters != null) {
            try {
                json = toDeterministicString(jsonParameters);
            } catch (JSONException e) {
                throw new RuntimeException(e.getMessage());
            }
        } else {
            json = "";
        }

        // Include the session token in the cache in order to avoid mixing permissions.
        if (sessionToken != null) {
            json += sessionToken;
        }

        return String.format(
                "ParseRESTCommand.%s.%s.%s",
                method.toString(),
                ParseDigestUtils.md5(httpPath),
                ParseDigestUtils.md5(json)
        );
    }

    public JSONObject toJSONObject() {
        JSONObject jsonObject = new JSONObject();
        try {
            if (httpPath != null) {
                jsonObject.put("httpPath", httpPath);
            }
            jsonObject.put("httpMethod", method.toString());
            if (jsonParameters != null) {
                jsonObject.put("parameters", jsonParameters);
            }
            if (sessionToken != null) {
                jsonObject.put("sessionToken", sessionToken);
            }
            if (localId != null) {
                jsonObject.put("localId", localId);
            }
        } catch (JSONException e) {
            throw new RuntimeException(e.getMessage());
        }
        return jsonObject;
    }

    public String getSessionToken() {
        return sessionToken;
    }

    public String getOperationSetUUID() {
        return operationSetUUID;
    }

    /* package */ void setOperationSetUUID(String operationSetUUID) {
        this.operationSetUUID = operationSetUUID;
    }

    public String getLocalId() {
        return localId;
    }

    public void setLocalId(String localId) {
        this.localId = localId;
    }

    /**
     * If this was the second save on a new object while offline, then its objectId wasn't yet set
     * when the command was created, so it would have been considered a "create". But if the first
     * save succeeded, then there is an objectId now, and it will be mapped to the localId for this
     * command's result. If so, change the "create" operation to an "update", and add the objectId to
     * the command.
     */
    private void maybeChangeServerOperation() {
        if (localId != null) {
            String objectId = getLocalIdManager().getObjectId(localId);
            if (objectId != null) {
                localId = null;
                httpPath += String.format("/%s", objectId);
                url = createUrl(httpPath);

                if (httpPath.startsWith("classes") && method == ParseHttpRequest.Method.POST) {
                    method = ParseHttpRequest.Method.PUT;
                }
            }
        }
    }

    public void resolveLocalIds() {
        try {
            ArrayList<JSONObject> localPointers = new ArrayList<>();
            getLocalPointersIn(jsonParameters, localPointers);
            for (JSONObject pointer : localPointers) {
                String localId = (String) pointer.get("localId");
                String objectId = getLocalIdManager().getObjectId(localId);
                if (objectId == null) {
                    throw new IllegalStateException(
                            "Tried to serialize a command referencing a new, unsaved object.");
                }
                pointer.put("objectId", objectId);
                pointer.remove("localId");
            }
            maybeChangeServerOperation();
        } catch (JSONException e) {
            // Well, nothing to do here...
        }
    }

    /**
     * Finds all of the local ids in this command and increments their retain counts in the on-disk
     * store. This should be called immediately before serializing the command to disk, so that we
     * know we might need to resolve these local ids at some point in the future.
     */
    public void retainLocalIds() {
        if (localId != null) {
            getLocalIdManager().retainLocalIdOnDisk(localId);
        }

        try {
            ArrayList<JSONObject> localPointers = new ArrayList<>();
            getLocalPointersIn(jsonParameters, localPointers);
            for (JSONObject pointer : localPointers) {
                String localId = (String) pointer.get("localId");
                getLocalIdManager().retainLocalIdOnDisk(localId);
            }
        } catch (JSONException e) {
            // Well, nothing to do here...
        }
    }

    /**
     * Finds all of the local ids in this command and decrements their retain counts in the on-disk
     * store. This should be called when removing a serialized command from the disk, when we know
     * that we will never need to resolve these local ids for this command again in the future.
     */
    public void releaseLocalIds() {
        if (localId != null) {
            getLocalIdManager().releaseLocalIdOnDisk(localId);
        }
        try {
            ArrayList<JSONObject> localPointers = new ArrayList<>();
            getLocalPointersIn(jsonParameters, localPointers);
            for (JSONObject pointer : localPointers) {
                String localId = (String) pointer.get("localId");
                getLocalIdManager().releaseLocalIdOnDisk(localId);
            }
        } catch (JSONException e) {
            // Well, nothing to do here...
        }
    }

    /* package */ static abstract class Init<T extends Init<T>> {
        public String masterKey;
        private String sessionToken;
        private String installationId;
        private ParseHttpRequest.Method method = ParseHttpRequest.Method.GET;
        private String httpPath;
        private JSONObject jsonParameters;

        private String operationSetUUID;
        private String localId;

        /* package */
        abstract T self();

        public T sessionToken(String sessionToken) {
            this.sessionToken = sessionToken;
            return self();
        }

        public T installationId(String installationId) {
            this.installationId = installationId;
            return self();
        }

        public T masterKey(String masterKey) {
            this.masterKey = masterKey;
            return self();
        }

        public T method(ParseHttpRequest.Method method) {
            this.method = method;
            return self();
        }

        public T httpPath(String httpPath) {
            this.httpPath = httpPath;
            return self();
        }

        public T jsonParameters(JSONObject jsonParameters) {
            this.jsonParameters = jsonParameters;
            return self();
        }

        public T operationSetUUID(String operationSetUUID) {
            this.operationSetUUID = operationSetUUID;
            return self();
        }

        public T localId(String localId) {
            this.localId = localId;
            return self();
        }
    }

    public static class Builder extends Init<Builder> {
        @Override
            /* package */ Builder self() {
            return this;
        }

        public ParseRESTCommand build() {
            return new ParseRESTCommand(this);
        }
    }
}

File: parse/src/main/java/com/parse/ParseRESTCommand.kt
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse

import com.parse.NoObjectsEncoder.Companion.get
import com.parse.boltsinternal.Task
import com.parse.http.ParseHttpBody
import com.parse.http.ParseHttpRequest
import com.parse.http.ParseHttpResponse
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.json.JSONStringer
import java.io.IOException
import java.io.InputStream
import java.net.MalformedURLException
import java.net.URL
import java.util.*

/**
 * A helper object to send requests to the server.
 */
internal open class ParseRESTCommand : ParseRequest<JSONObject?> {
    @JvmField
    val jsonParameters: JSONObject?

    // Headers
    val sessionToken: String?
    var masterKey: String? = null

    /* package */
    @JvmField
    var httpPath: String?
    var installationId: String? = null
    var operationSetUUID: String? = null
    var localId: String?

    constructor(
        httpPath: String?,
        httpMethod: ParseHttpRequest.Method?,
        parameters: Map<String?, *>?,
        sessionToken: String?
    ) : this(
        httpPath,
        httpMethod,
        if (parameters != null) get().encode(parameters) as JSONObject else null,
        sessionToken
    )

    constructor(
        httpPath: String?,
        httpMethod: ParseHttpRequest.Method?,
        jsonParameters: JSONObject?,
        sessionToken: String?
    ) : this(httpPath, httpMethod, jsonParameters, null, sessionToken) {
    }

    private constructor(
        httpPath: String?,
        httpMethod: ParseHttpRequest.Method?,
        jsonParameters: JSONObject?,
        localId: String?, sessionToken: String?
    ) : super(httpMethod!!, createUrl(httpPath)) {
        this.httpPath = httpPath
        this.jsonParameters = jsonParameters
        this.localId = localId
        this.sessionToken = sessionToken
    }

    constructor(builder: Init<*>) : super(builder.method, createUrl(builder.httpPath)) {
        sessionToken = builder.sessionToken
        installationId = builder.installationId
        masterKey = builder.masterKey
        httpPath = builder.httpPath
        jsonParameters = builder.jsonParameters
        operationSetUUID = builder.operationSetUUID
        localId = builder.localId
    }

    protected open fun addAdditionalHeaders(requestBuilder: ParseHttpRequest.Builder) {
        if (installationId != null) {
            requestBuilder.addHeader(HEADER_INSTALLATION_ID, installationId)
        }
        if (sessionToken != null) {
            requestBuilder.addHeader(HEADER_SESSION_TOKEN, sessionToken)
        }
        if (masterKey != null) {
            requestBuilder.addHeader(HEADER_MASTER_KEY, masterKey)
        }
    }

    override fun newRequest(
        method: ParseHttpRequest.Method,
        url: String?,
        uploadProgressCallback: ProgressCallback?
    ): ParseHttpRequest {
        val request: ParseHttpRequest = if (jsonParameters != null && method != ParseHttpRequest.Method.POST && method != ParseHttpRequest.Method.PUT) {
                // The request URI may be too long to include parameters in the URI.
                // To avoid this problem we send the parameters in a POST request json-encoded body
                // and add a http method override parameter in newBody.
                super.newRequest(ParseHttpRequest.Method.POST, url, uploadProgressCallback)
            } else {
                super.newRequest(method, url, uploadProgressCallback)
            }
        val requestBuilder = ParseHttpRequest.Builder(request)
        addAdditionalHeaders(requestBuilder)
        return requestBuilder.build()
    }

    public override fun newBody(uploadProgressCallback: ProgressCallback?): ParseHttpBody? {
        if (jsonParameters == null) {
            val message = String.format(
                "Trying to execute a %s command without body parameters.",
                method.toString()
            )
            throw IllegalArgumentException(message)
        }
        return try {
            var parameters: JSONObject = jsonParameters
            if (method == ParseHttpRequest.Method.GET ||
                method == ParseHttpRequest.Method.DELETE
            ) {
                // The request URI may be too long to include parameters in the URI.
                // To avoid this problem we send the parameters in a POST request json-encoded body
                // and add a http method override parameter.
                parameters = JSONObject(jsonParameters.toString())
                parameters.put(PARAMETER_METHOD_OVERRIDE, method.toString())
            }
            ParseByteArrayHttpBody(parameters.toString(), "application/json")
        } catch (e: Exception) {
            throw RuntimeException(e.message)
        }
    }

    override fun executeAsync(
        client: ParseHttpClient,
        uploadProgressCallback: ProgressCallback?,
        downloadProgressCallback: ProgressCallback?,
        cancellationToken: Task<Void>?
    ): Task<JSONObject?> {
        resolveLocalIds()
        return super.executeAsync(
            client, uploadProgressCallback, downloadProgressCallback, cancellationToken
        )
    }

    public override fun onResponseAsync(
        response: ParseHttpResponse?,
        downloadProgressCallback: ProgressCallback?
    ): Task<JSONObject?>? {
        val content: String
        var responseStream: InputStream? = null
        try {
            responseStream = response!!.content
            content = String(ParseIOUtils.toByteArray(responseStream))
        } catch (e: IOException) {
            return Task.forError(e)
        } finally {
            ParseIOUtils.closeQuietly(responseStream)
        }

        // We need to check for errors differently in /1/ than /2/ since object data in /2/ was
        // encapsulated in "data" and everything was 200, but /2/ everything is in the root JSON,
        // but errors are status 4XX.
        // See https://quip.com/4pbbA9HbOPjQ
        val statusCode = response!!.statusCode
        if (statusCode in 200..599) { // Assume 3XX is handled by http library
            val json: JSONObject
            return try {
                json = JSONObject(content)
                if (statusCode in 400..499) { // 4XX
                    return Task.forError(
                        newPermanentException(
                            json.optInt("code"),
                            json.optString("error")
                        )
                    )
                } else if (statusCode >= 500) { // 5XX
                    return Task.forError(
                        newTemporaryException(
                            json.optInt("code"),
                            json.optString("error")
                        )
                    )
                }
                Task.forResult(json)
            } catch (e: JSONException) {
                Task.forError(newTemporaryException("bad json response", e))
            }
        }
        return Task.forError(newPermanentException(ParseException.OTHER_CAUSE, content))
    }// Include the session token in the cache in order to avoid mixing permissions.

    // Creates a somewhat-readable string that uniquely identifies this command.
    val cacheKey: String
        get() {
            var json: String? = if (jsonParameters != null) {
                try {
                    toDeterministicString(jsonParameters)
                } catch (e: JSONException) {
                    throw RuntimeException(e.message)
                }
            } else {
                ""
            }

            // Include the session token in the cache in order to avoid mixing permissions.
            if (sessionToken != null) {
                json += sessionToken
            }
            return String.format(
                "ParseRESTCommand.%s.%s.%s",
                method.toString(),
                ParseDigestUtils.md5(httpPath),
                ParseDigestUtils.md5(json)
            )
        }

    fun toJSONObject(): JSONObject {
        val jsonObject = JSONObject()
        try {
            if (httpPath != null) {
                jsonObject.put("httpPath", httpPath)
            }
            jsonObject.put("httpMethod", method.toString())
            if (jsonParameters != null) {
                jsonObject.put("parameters", jsonParameters)
            }
            if (sessionToken != null) {
                jsonObject.put("sessionToken", sessionToken)
            }
            if (localId != null) {
                jsonObject.put("localId", localId)
            }
        } catch (e: JSONException) {
            throw RuntimeException(e.message)
        }
        return jsonObject
    }

    /**
     * If this was the second save on a new object while offline, then its objectId wasn't yet set
     * when the command was created, so it would have been considered a "create". But if the first
     * save succeeded, then there is an objectId now, and it will be mapped to the localId for this
     * command's result. If so, change the "create" operation to an "update", and add the objectId to
     * the command.
     */
    private fun maybeChangeServerOperation() {
        if (localId != null) {
            val objectId = localIdManager.getObjectId(localId!!)
            if (objectId != null) {
                localId = null
                httpPath += String.format("/%s", objectId)
                url = createUrl(httpPath)
                if (httpPath!!.startsWith("classes") && method == ParseHttpRequest.Method.POST) {
                    method = ParseHttpRequest.Method.PUT
                }
            }
        }
    }

    fun resolveLocalIds() {
        try {
            val localPointers = ArrayList<JSONObject>()
            getLocalPointersIn(jsonParameters, localPointers)
            for (pointer in localPointers) {
                val localId = pointer["localId"] as String
                val objectId = localIdManager.getObjectId(localId)
                    ?: throw IllegalStateException(
                        "Tried to serialize a command referencing a new, unsaved object."
                    )
                pointer.put("objectId", objectId)
                pointer.remove("localId")
            }
            maybeChangeServerOperation()
        } catch (e: JSONException) {
            // Well, nothing to do here...
        }
    }

    /**
     * Finds all of the local ids in this command and increments their retain counts in the on-disk
     * store. This should be called immediately before serializing the command to disk, so that we
     * know we might need to resolve these local ids at some point in the future.
     */
    fun retainLocalIds() {
        if (localId != null) {
            localIdManager.retainLocalIdOnDisk(localId!!)
        }
        try {
            val localPointers = ArrayList<JSONObject>()
            getLocalPointersIn(jsonParameters, localPointers)
            for (pointer in localPointers) {
                val localId = pointer["localId"] as String
                localIdManager.retainLocalIdOnDisk(localId)
            }
        } catch (e: JSONException) {
            // Well, nothing to do here...
        }
    }

    /**
     * Finds all of the local ids in this command and decrements their retain counts in the on-disk
     * store. This should be called when removing a serialized command from the disk, when we know
     * that we will never need to resolve these local ids for this command again in the future.
     */
    fun releaseLocalIds() {
        if (localId != null) {
            localIdManager.releaseLocalIdOnDisk(localId!!)
        }
        try {
            val localPointers = ArrayList<JSONObject>()
            getLocalPointersIn(jsonParameters, localPointers)
            for (pointer in localPointers) {
                val localId = pointer["localId"] as String
                localIdManager.releaseLocalIdOnDisk(localId)
            }
        } catch (e: JSONException) {
            // Well, nothing to do here...
        }
    }

    /* package */
    internal abstract class Init<T : Init<T>?> {
        var masterKey: String? = null
        var sessionToken: String? = null
        var installationId: String? = null
        var method = ParseHttpRequest.Method.GET
        var httpPath: String? = null
        var jsonParameters: JSONObject? = null
        var operationSetUUID: String? = null
        var localId: String? = null

        /* package */
        abstract fun self(): T
        fun sessionToken(sessionToken: String?): T {
            this.sessionToken = sessionToken
            return self()
        }

        fun installationId(installationId: String?): T {
            this.installationId = installationId
            return self()
        }

        fun masterKey(masterKey: String?): T {
            this.masterKey = masterKey
            return self()
        }

        fun method(method: ParseHttpRequest.Method): T {
            this.method = method
            return self()
        }

        fun httpPath(httpPath: String?): T {
            this.httpPath = httpPath
            return self()
        }

        fun jsonParameters(jsonParameters: JSONObject?): T {
            this.jsonParameters = jsonParameters
            return self()
        }

        fun operationSetUUID(operationSetUUID: String?): T {
            this.operationSetUUID = operationSetUUID
            return self()
        }

        fun localId(localId: String?): T {
            this.localId = localId
            return self()
        }
    }

    class Builder : Init<Builder>() {
        override fun  /* package */self(): Builder {
            return this
        }

        fun build(): ParseRESTCommand {
            return ParseRESTCommand(this)
        }
    }

    companion object {
        /* package */
        const val HEADER_APPLICATION_ID = "X-Parse-Application-Id"

        /* package */
        const val HEADER_CLIENT_KEY = "X-Parse-Client-Key"

        /* package */
        const val HEADER_APP_BUILD_VERSION = "X-Parse-App-Build-Version"

        /* package */
        const val HEADER_APP_DISPLAY_VERSION = "X-Parse-App-Display-Version"

        /* package */
        const val HEADER_OS_VERSION = "X-Parse-OS-Version"

        /* package */
        const val HEADER_INSTALLATION_ID = "X-Parse-Installation-Id"

        /* package */
        const val USER_AGENT = "User-Agent"
        private const val HEADER_SESSION_TOKEN = "X-Parse-Session-Token"
        private const val HEADER_MASTER_KEY = "X-Parse-Master-Key"
        private const val PARAMETER_METHOD_OVERRIDE = "_method"

        // Set via Parse.initialize(Configuration)
        /* package */
        @JvmField
        var server: URL? = null
        private val localIdManager: LocalIdManager
            get() = ParseCorePlugins.getInstance().localIdManager

        @JvmStatic
        fun fromJSONObject(jsonObject: JSONObject): ParseRESTCommand {
            val httpPath = jsonObject.optString("httpPath")
            val httpMethod = ParseHttpRequest.Method.fromString(jsonObject.optString("httpMethod"))
            val sessionToken = jsonObject.optString("sessionToken", null)
            val localId = jsonObject.optString("localId", null)
            val jsonParameters = jsonObject.optJSONObject("parameters")
            return ParseRESTCommand(httpPath, httpMethod, jsonParameters, localId, sessionToken)
        }

        private fun createUrl(httpPath: String?): String {
            // We send all parameters for GET/HEAD/DELETE requests in a post body,
            // so no need to worry about query parameters here.
            return if (httpPath == null) {
                server.toString()
            } else try {
                URL(server, httpPath).toString()
            } catch (ex: MalformedURLException) {
                throw RuntimeException(ex)
            }
        }

        // Encodes the object to JSON, but ensures that JSONObjects
        // and nested JSONObjects are encoded with keys in alphabetical order.
        @JvmStatic
        @Throws(JSONException::class)
        fun toDeterministicString(o: Any): String {
            val stringer = JSONStringer()
            addToStringer(stringer, o)
            return stringer.toString()
        }

        // Uses the provided JSONStringer to encode this object to JSON, but ensures that JSONObjects and
        // nested JSONObjects are encoded with keys in alphabetical order.
        @Throws(JSONException::class)
        private fun addToStringer(stringer: JSONStringer, o: Any) {
            if (o is JSONObject) {
                stringer.`object`()
                val `object` = o
                val keyIterator = `object`.keys()
                val keys = ArrayList<String>()
                while (keyIterator.hasNext()) {
                    keys.add(keyIterator.next())
                }
                keys.sort()
                for (key in keys) {
                    stringer.key(key)
                    addToStringer(stringer, `object`.opt(key))
                }
                stringer.endObject()
                return
            }
            if (o is JSONArray) {
                val array = o
                stringer.array()
                for (i in 0 until array.length()) {
                    addToStringer(stringer, array[i])
                }
                stringer.endArray()
                return
            }
            stringer.value(o)
        }

        /* package */
        @JvmStatic
        fun isValidCommandJSONObject(jsonObject: JSONObject): Boolean {
            return jsonObject.has("httpPath")
        }

        // This function checks whether a json object is a valid /2 ParseCommand json.
        /* package */
        @JvmStatic
        fun isValidOldFormatCommandJSONObject(jsonObject: JSONObject): Boolean {
            return jsonObject.has("op")
        }

        @Throws(JSONException::class)
        protected fun getLocalPointersIn(container: Any?, localPointers: ArrayList<JSONObject>) {
            if (container is JSONObject) {
                if ("Pointer" == container.opt("__type") && container.has("localId")) {
                    localPointers.add(container)
                    return
                }
                val keyIterator = container.keys()
                while (keyIterator.hasNext()) {
                    val key = keyIterator.next()
                    getLocalPointersIn(container[key], localPointers)
                }
            }
            if (container is JSONArray) {
                for (i in 0 until container.length()) {
                    getLocalPointersIn(container[i], localPointers)
                }
            }
        }
    }
}